### PR TITLE
General improvements

### PR DESCRIPTION
**Pull request by macgyver603**
The goal of this PR is to improve the asset by generalizing code between processors where possible and externalizing it to separate modules within the asset. Additionally, the HDFS asset bundle will be added to share the benefits of the generalized functions as well as allow the `chunked-file-reader` to be added as a module in this asset instead of the Teraslice monorepo.

Tasks:
- [x] Generalize S3/file exporter code  
- [x] Generalize S3/file reader code  
- [x] Import HDFS asset bundle
- [x] Import `chunked-file-reader`

### CODE REVIEW

**Comment by macgyver603** [comment]
This is ready for review now

File: asset/_lib/chunked-file-reader/index.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+// This function will grab the chunk of data specified by the slice plus an
+// extra margin if the slice does not end with the delimiter.
+function getChunk(readerClient, slice, opConfig, logger, metadata) {
+    const delimiter = opConfig.line_delimiter;
+
+    async function getMargin(offset, length) {
+        let margin = '';
+        while (margin.indexOf(delimiter) === -1) {
+            // reader clients must return false-y when nothing more to read.
+            const chunk = await readerClient(offset, length); // eslint-disable-line no-await-in-loop, max-len
```
Comment (peterdemartini) [inline asset/_lib/chunked-file-reader/index.js]: We've disabled this eslint rule:

```suggestion
            const chunk = await readerClient(offset, length);
```

File: asset/_lib/chunked-file-reader/formatters.js [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        output: 'json'
+    }, opConfig.extra_args);
+    let foundHeader = false;
+    const data = _toRecords(incomingData, opConfig.line_delimiter, slice);
+    return Promise.map(data, async (record) => {
+        try {
+            let parsedLine = await csvToJson(csvParams)
+                .fromString(record).then((parsedData) => parsedData[0]);
+            // csvToJson trim applied inconsistently so implemented this function
+            _.keys(parsedLine).forEach((key) => {
```
Comment (peterdemartini) [inline asset/_lib/chunked-file-reader/formatters.js]:
```suggestion
            Object.keys(parsedLine).forEach((key) => {
```

File: asset/_lib/chunked-file-reader/formatters.js [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                parsedLine[key] = parsedLine[key].trim();
+            });
+            // Check for header row. Assumes there would only be one header row in a slice
+            if (opConfig.remove_header && !foundHeader
+                && _.keys(parsedLine).sort().join() === _.values(parsedLine).sort().join()) {
```
Comment (peterdemartini) [inline asset/_lib/chunked-file-reader/formatters.js]:
```suggestion
                && Object.keys(parsedLine).sort().join() === Object.values(parsedLine).sort().join()) {
```

File: asset/_lib/chunked-file-reader/formatters.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    if (rawData.endsWith(delimiter)) {
+        // Get rid of last character if it is the delimiter since that will
+        // just result in an empty record.
+        outputData = rawData.slice(0, -delimiter.length);
+    }
+
+    if (slice.offset === 0) {
+        return outputData.split(delimiter);
+    }
+    return outputData.split(delimiter).splice(1);
```
[the remark this answers is not in the document]
Comment (macgyver603) [reply asset/_lib/chunked-file-reader/formatters.js]: fixed

File: asset/_lib/parser.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        const json = () => `${JSON.stringify(slice)}${opConfig.line_delimiter}`;
+        const formats = {
+            csv,
+            tsv: csv,
+            raw,
+            ldjson,
+            json,
+            default: () => new Error(`Unsupported output format "${opConfig.format}"`)
+        };
+        return (formats[format] || formats.default)();
```
[the remark this answers is not in the document]
Comment (macgyver603) [reply asset/_lib/parser.js]: fixed

File: asset/_lib/slice.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
+'use strict';
+
+const { getOffsets } = require('@terascope/chunked-file-reader');
```
[the remark this answers is not in the document]
Comment (macgyver603) [reply asset/_lib/slice.js]: fixed

File: asset/_lib/stringify.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
+'use strict';
+
+function stringify(record, fields) {
+    let serializedRecord = '{';
+    fields.forEach((field) => {
+        // Can't just check for `record[field]` since `null`, `undefined`, and `0` will drop
+        // fields
+        if (Object.keys(record).includes(field)) {
```
[the remark this answers is not in the document]
Comment (macgyver603) [reply asset/_lib/stringify.js]: The custom JSON stringify has been abandoned for the builtin stringify

File: asset/_lib/parser.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    };
+
+    const outStr = parseData(opConfig.format);
+
+    if (outStr instanceof Error) throw outStr;
```
[the remark this answers is not in the document]
Comment (macgyver603) [reply asset/_lib/parser.js]: fixed

File: test/lib/lib-spec.js [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+describe('File asset supporting library', () => {
+    describe('parser module', () => {
+        it('errors with invalid formats.', async () => {
+            let error;
+            try {
+                await parseForFile([{}], { format: 'invalid' });
+            } catch (e) {
+                error = e;
+            }
+            expect(error).toEqual(new Error('Unsupported output format "invalid"'));
```
[the remark this answers is not in the document]
Comment (macgyver603) [reply test/lib/lib-spec.js]: fixed

File: asset/_lib/parser.js [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+            slice.forEach((record) => {
+                outStr = `${outStr}${record.data}${opConfig.line_delimiter}`;
+            });
+            return outStr;
+        };
+        const ldjson = () => {
+            let outStr = '';
+            if (opConfig.fields.length > 0) {
+                slice.forEach((record) => {
+                    outStr = `${outStr}${stringify(record, opConfig.fields)}${opConfig.line_delimiter}`;
```
[the remark this answers is not in the document]
Comment (macgyver603) [reply asset/_lib/parser.js]: fixed

File: asset/_lib/chunked-file-reader/index.js (hunk 1 above)
Comment (macgyver603) [reply asset/_lib/chunked-file-reader/index.js]: fixed

File: asset/_lib/chunked-file-reader/formatters.js (hunk 2 above)
Comment (macgyver603) [reply asset/_lib/chunked-file-reader/formatters.js]: fixed

File: asset/_lib/chunked-file-reader/formatters.js (hunk 3 above)
Comment (macgyver603) [reply asset/_lib/chunked-file-reader/formatters.js]: fixed

File: asset/file_timeseries_router/processor.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+                '/',
+                opConfig.prefix
+            )
+        );
+
+        return record;
+    }
+
+    async onBatch(slice) {
+        return Promise.all(slice, (record) => this.addPath(record));
```
Comment (peterdemartini) [inline asset/file_timeseries_router/processor.js]:
```suggestion
        return Promise.all(slice.map((record) => this.addPath(record)));
```

**Pull request merged**